src/main.cpp

```diff
@@ -866,12 +919,36 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
 {
     // Basic checks that don't depend on any context
 
-    // Check transaction version
-    if (tx.nVersion < MIN_TX_VERSION) {
+    // Pre-Overwinter accepts version >=1 transaction. Overwinter only accepts version 3 transactions.
+    // To enable soft-forks of the transaction format, define a range of versions which can be accepted.
+    if (tx.fOverwintered) {
+        if (tx.nVersion < CTransaction::OVERWINTER_MIN_CURRENT_VERSION) {
+            return state.DoS(100, error("CheckTransaction(): overwinter version too low"),
+                REJECT_INVALID, "bad-tx-overwinter-version-too-low");
+        }
+        else if (tx.nVersion > CTransaction::OVERWINTER_MAX_CURRENT_VERSION ) {
+            return state.DoS(100, error("CheckTransaction(): overwinter version too high"),
+                REJECT_INVALID, "bad-tx-overwinter-version-too-high");
+        }
+    }
+
+    if (!tx.fOverwintered && tx.nVersion < MIN_TX_VERSION) {
         return state.DoS(100, error("CheckTransaction(): version too low"),
                          REJECT_INVALID, "bad-txns-version-too-low");
     }
 
+    // Check Overwintered tx version group id matches what Zcash has assigned to v3 tx format version.
+    if (tx.fOverwintered && tx.nVersion == 3 && tx.nVersionGroupId != OVERWINTER_VERSION_GROUP_ID) {
```

Comment:
I know ``OVERWINTER_MAX_CURRENT_VERSION`` is set to 3 so it doesn't matter for this upgrade, but it still seems semantically wrong for this to be conditional on ``tx.nVersion == 3``. The version group ID should be a recognized value, for any version in the range ``${UPGRADE}_MIN_CURRENT_VERSION`` to ``${UPGRADE}_MAX_CURRENT_VERSION``. Since this code is likely to be copied for future upgrades, I'd prefer the more restrictive condition (i.e. just ``(tx.fOverwintered && tx.nVersionGroupId != OVERWINTER_VERSION_GROUP_ID)``).